core/jvm/src/main/scala-2.13/reftree/core/HackedCollectionInstances.scala

```diff
@@ -0,0 +1,211 @@
+    bitMapIndexedSetNodeRefTree[A](value.rootNode).rename("HashSet")
+  }
+
+  private def bitMapIndexedSetNodeRefTree[A: ToRefTree](node: BitmapIndexedSetNode[A]): RefTree.Ref = {
+    val hash =
+      RefTree.Val.formatted(node.cachedJavaKeySetHashCode)(_.toHexString)
+        .toField.withName("hash")
+    val size = node.size.refTree.toField.withName("size")
+    val nodeArity = node.nodeArity.refTree.toField.withName("nodeArity")
+    val payloadArity = node.payloadArity.refTree.toField.withName("payloadArity")
+
+    val content =
+      RefTree.Ref(
+          node.content,
+          hashSetArrayRefTree[A](node.content, node.nodeArity)
+        )
+        .rename("Array")
+        .toField
+
+    RefTree.Ref(
+      node,
+      Seq(hash, size, nodeArity, payloadArity, content)
+    )
+  }
+
+  private def hashSetArrayRefTree[A: ToRefTree](content: Array[Any], arity: Int): Seq[RefTree.Ref.Field] =
+    if (arity > 0) {
+      content.toVector.map { x =>
+        if (x.isInstanceOf[BitmapIndexedSetNode[_]])
+          bitMapIndexedSetNodeRefTree[A](x.asInstanceOf[BitmapIndexedSetNode[A]])
+        else
+          x.asInstanceOf[A].refTree
+      }.map(_.toField)
+    } else
+      content.toVector.map(_.asInstanceOf[A].refTree.toField)
+
+  implicit def `HashMap RefTree`[A: ToRefTree, B: ToRefTree]: ToRefTree[HashMap[A, B]] =
+    ToRefTree[HashMap[A, B]] { value =>
+      bitmapIndexedMapNodeRefTree[A, B](value.rootNode).rename("HashMap")
+    }
+
+  private def bitmapIndexedMapNodeRefTree[A: ToRefTree, B: ToRefTree](
+    node: BitmapIndexedMapNode[A, B]
+  ): RefTree.Ref = {
+    val hash =
+      RefTree.Val.formatted(node.cachedJavaKeySetHashCode)(_.toHexString)
+        .toField.withName("hash")
+    val size = node.size.refTree.toField.withName("size")
+    val nodeArity = node.nodeArity.refTree.toField.withName("nodeArity")
+    val payloadArity = node.payloadArity.refTree.toField.withName("payloadArity")
+
+    val content =
+      RefTree.Ref(
+        node.content,
+        hashMapArrayRefTree[A, B](node.content, node.nodeArity)
+      )
+      .rename("Array")
+      .toField
+
+    RefTree.Ref(
+      node,
+      Seq(hash, size, nodeArity, payloadArity, content)
+    )
+  }
+
+  private def hashMapArrayRefTree[A: ToRefTree, B: ToRefTree](
+    content: Array[Any],
+    arity: Int
+  ): Seq[RefTree.Ref.Field] =
+  if (arity > 0) {
+    val mapNodesBuilder = Vector.newBuilder[RefTree.Ref]
+    val keysBuilder = Vector.newBuilder[A]
+    val valuesBuilder = Vector.newBuilder[B]
+
+    content.toVector.zipWithIndex.foreach {
+      case (el, _) if el.isInstanceOf[BitmapIndexedMapNode[_, _]] =>
+        mapNodesBuilder += bitmapIndexedMapNodeRefTree[A, B](el.asInstanceOf[BitmapIndexedMapNode[A, B]])
+      case (el, i) if i % 2 == 0 =>
+        keysBuilder += el.asInstanceOf[A]
+      case (el, _) =>
+        valuesBuilder += el.asInstanceOf[B]
+    }
+
+    (keysBuilder.result().zip(valuesBuilder.result()).map(_.refTree) ++
```

Comment:
@danicheg unlike the HashSet instance, this one does not maintain the order of nodes. I think that might get in the way of understanding how the data structure works. Per https://michael.steindorfer.name/publications/oopsla15.pdf, it’s possible that nodes come before values:
<img width="329" alt="image" src="https://github.com/stanch/reftree/assets/2670454/937a7d70-c93d-436d-a8a3-e25ac12bdb9e">

It seems you’ve done this on purpose (since just mapping stuff to a single vector of reftrees would be much simpler). Could you elaborate on your logic?

Reply:
Reading further into the paper, after all it seems that stuff is reordered for nodes to come last? I guess otherwise the `i % 2` logic would not work anyway :D
<img width="1073" alt="image" src="https://github.com/stanch/reftree/assets/2670454/f37345b1-be3d-4875-a677-3c7c91e150c3">

Reply:
> unlike the HashSet instance, this one does not maintain the order of nodes.

In Scala 2.13 Collections Library, `HashSet` does not preserve the original order of inserted items. Only `LinkedHashSet` preserve them. Similarly, `HashMap` also doesn't preserve the order, and only `LinkedHashMap` does. Keeping that in mind, I haven't been pursued the ordering much.

Reply:
I meant not the order of elements, but the order of values and nodes in compressed arrays, as shown on the pictures from the paper

Reply:
Hmmm, I'm struggling to definitively say on the accuracy of Scala's implementation compared to the original paper. I took an easier approach and performed some basic integration testing for each instance of the collections. It worked fine, but surely, there could be corner cases where things might behave differently 🤷🏻

Reply:
Ah, I remember where I've seen that pattern, where values come right after values in the resulting `content: Array[Any]` — https://github.com/scala/scala/blob/v2.13.14/src/library/scala/collection/immutable/HashMap.scala#L1532-L1533.

Reply:
But yeah, as I mentioned above, there could be corner cases that break the common norm.

Reply:
The invariants in the code seem to suggest that your reading is correct — the array is organized such that keys and values come first, and nodes after that: https://github.com/scala/scala/blob/v2.13.14/src/library/scala/collection/immutable/HashMap.scala#L642

Reply:
The only remaining doubt I have is whether the “cached Java hash code” should be shown in the tree. It seems like an optimization for equality checks, rather than what is used to actually address the nodes...

Reply:
> “cached Java hash code” should be shown in the tree


My reasoning was that someone might want to have some IDs for the nodes to simplify explaining things while giving a talk 😄 But yeah, it's something auxiliary indeed.

Reply:
I’ve replaced this with `dataMap` and `nodeMap` shown in binary — similar to how 2.12 HashSet/HasMap had the bitmap shown.